kernel/kernel-api/src/main/java/io/delta/kernel/transaction/DataLayoutSpec.java

```diff
@@ -0,0 +1,153 @@
+   */
+  public static DataLayoutSpec unpartitioned() {
+    return new DataLayoutSpec(null, null);
+  }
+
+  /**
+   * Creates a data layout specification for a partitioned table.
+   *
+   * @param partitionColumns The columns to partition by. Cannot be null or empty. Only top-level
+   *     columns are supported for partitioning.
+   * @return A new {@link DataLayoutSpec} for a partitioned table.
+   */
+  public static DataLayoutSpec partitioned(List<Column> partitionColumns) {
+    checkArgument(
+        partitionColumns != null && !partitionColumns.isEmpty(),
+        "Partition columns cannot be null or empty");
+    checkArgument(
+        partitionColumns.stream().allMatch(col -> col.getNames().length == 1),
+        "Partition columns must be only top-level columns");
+    return new DataLayoutSpec(partitionColumns, null);
+  }
+
+  /**
+   * Creates a data layout specification for a clustered table.
+   *
+   * @param clusteringColumns The columns to cluster by. Cannot be null, but can be empty to
+   *     indicate clustering is enabled without specific column definitions.
+   * @return A new {@link DataLayoutSpec} for a clustered table.
+   */
+  public static DataLayoutSpec clustered(List<Column> clusteringColumns) {
+    checkArgument(
+        clusteringColumns != null, "Clustering columns cannot be null (but can be empty)");
+    return new DataLayoutSpec(null, clusteringColumns);
+  }
+
+  private final List<Column> partitionColumns;
+  private final List<Column> clusteringColumns;
+
+  private DataLayoutSpec(List<Column> partitionColumns, List<Column> clusteringColumns) {
+    if (partitionColumns != null && clusteringColumns != null) {
+      throw new IllegalArgumentException("Cannot specify both partition and clustering columns");
+    }
+    if (partitionColumns != null && partitionColumns.isEmpty()) {
+      throw new IllegalArgumentException("Partition columns cannot be empty");
+    }
+    this.partitionColumns = partitionColumns;
+    this.clusteringColumns = clusteringColumns;
+  }
+
+  /**
+   * Returns true if this layout specification includes partitioning.
+   *
+   * <p>Partitioning requires non-empty partition columns. An empty list of partition columns is not
+   * considered valid partitioning.
+   */
+  public boolean hasPartitioning() {
+    return partitionColumns != null && !partitionColumns.isEmpty();
+  }
+
+  /**
+   * Returns true if this layout specification includes clustering.
+   *
+   * <p>Clustering can be enabled even with empty clustering columns, which indicates that
+   * clustering is enabled on the table but no specific columns are defined yet.
+   */
+  public boolean hasClustering() {
+    return clusteringColumns != null;
+  }
+
+  /**
+   * Returns true if this is an unpartitioned layout.
+   *
+   * <p>An unpartitioned layout has neither partitioning nor clustering enabled.
+   */
+  public boolean isUnpartitioned() {
```

Comment:
`isNoDataLayoutSpec` or something more like that? e.g. above you say `<li><b>No data layout spec:</b> No special data layout`

Reply:
e.g. `hasPartitioning` = false doesn't actually mean that `isUnpartitioned=true` which is a tad confusing
but if this was `isNoDataLayoutSpec` that makes it clearer it applies to both partitioning and clustering

Reply:
updated this in https://github.com/delta-io/delta/pull/5056